# Laboratorio 7: Clasificación 🤗

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>


### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Eduardo Moya
- Ayudantes: Nicolás Ojeda, Melanie Peña, Valentina Rojas

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Fabrizzio Pezzolla
- Nombre de alumno 2: Rodrigo Molina Ávila

### **Link de repositorio de GitHub:** [Insertar Repositorio](https://github.com/frpezzolla/lab_mds.git)

### Temas a tratar
- Clasificación en problemas desbalanceados
- Lightgbm y xgboost
- Pipelines

### Reglas:

- **Grupos de 2 personas**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias.
- Pueden usar cualquer matrial del curso que estimen conveniente.
- Código que no se pueda ejecutar, no será revisado.



### Objetivos principales del laboratorio
- Comprender cómo trabajar con problemas de clasificación con clases desbalanceadas.
- Aplicar los modelos lightgbm y xgboost.
- Practicar Pipelines

# Parte Teórica [12 puntos]




1. Explique cuál es la diferencia entre los datos de entrenamiento y validación. [1 punto]

2. Explique cuál es el principal desafío al trabajar problemas de clasificación con data no supervisada. [1 punto]

3. Explique en **sus palabras** qué es la matriz de confusión y para qué se utiliza. [1 puntos]

4. Escriba la fórmula de las siguientes métricas y explique con **sus palabras** cómo se interpretan. [1 punto cada uno]

  * Accuracy
  * Precision
  * Recall
  * F1 score

5. Explique qué métrica recomendaría para los siguientes contextos de clasificación. [1 punto cada uno]

  * Mantenimiento predictivo de fallas de maquinaria pesada en la industria minera.  
  * Detección de enfermedades altamente contagiosas.
  * Aprobación de créditos de alto riesgo.
  * Detección de crímenes.

6. Explique qué es la calibración de modelos y para qué se usa. [1 punto]


**Respuesta:**

1#

Los datos de entrenamiento son aquellos que se utilizan para ajustar el modelo, es decir, para que el modelo aprenda los patrones presentes en los datos, este se deja intacto durance el proceso. Los datos de validación, por otro lado, se utilizan para evaluar el rendimiento del modelo durante el proceso de entrenamiento y ajustar los hiperparámetros, asegurando que el modelo generalice bien a datos no vistos.

2#

El principal desafío al trabajar con problemas de clasificación con datos no supervisados es la falta de etiquetas que indiquen la clase a la que pertenece cada ejemplo. Esto dificulta la evaluación del rendimiento del modelo y la identificación de patrones relevantes en los datos.

3#

La matriz de confusión es una herramienta que permite evaluar el rendimiento de un modelo de clasificación. Se utiliza para comparar las predicciones del modelo con las verdaderas etiquetas de los datos, mostrando el número de verdaderos positivos, verdaderos negativos, falsos positivos y falsos negativos. Esto ayuda a entender mejor los errores del modelo y a calcular métricas de rendimiento como la precisión, el recall y el F1 score.

4#

* Accuracy: (TP + TN) / (TP + TN + FP + FN). La accuracy mide la proporción de predicciones correctas sobre el total de predicciones. Se interpreta como la capacidad del modelo para clasificar correctamente tanto las clases positivas como las negativas.
* Precision: TP / (TP + FP). La precision mide la proporción de verdaderos positivos sobre el total de predicciones positivas. Se interpreta como la capacidad del modelo para evitar falsos positivos.
* Recall: TP / (TP + FN). El recall mide la proporción de verdaderos positivos sobre el total de ejemplos positivos reales. Se interpreta como la capacidad del modelo para identificar correctamente las clases positivas.
* F1 score: 2 * (Precision * Recall) / (Precision + Recall). El F1 score es la media armónica de la precision y el recall. Se interpreta como un equilibrio entre la capacidad del modelo para evitar falsos positivos y su capacidad para identificar correctamente las clases positivas.

5#

* Mantenimiento predictivo de fallas de maquinaria pesada en la industria minera: Recall, ya que es crucial identificar todas las posibles fallas para evitar accidentes y costos elevados.
* Detección de enfermedades altamente contagiosas: Recall, porque es importante identificar todos los casos posibles para prevenir la propagación de la enfermedad.
* Aprobación de créditos de alto riesgo: Precision, para minimizar el riesgo de aprobar créditos a personas que no podrán pagarlos.
* Detección de crímenes: Precision, para evitar acusar a personas inocentes y asegurar que los verdaderos culpables sean identificados.

6#

La calibración de modelos es el proceso de ajustar las probabilidades de salida de un modelo para que reflejen mejor la realidad. Se utiliza para mejorar la interpretación de las probabilidades predichas y asegurar que las decisiones basadas en estas probabilidades sean más precisas y confiables.


# Parte práctica [48 puntos]

Tras el trágico despido de la mítica mascota de Maipú, Renacín decide adentrarse como consultor en el mercado futbolero, el cuál (para variar...) está cargado en especulaciones.

Como su principal tarea será asesorar a los directivos de los clubes sobre cuál jugador comprar y cuál no, Renacín desea generar modelos predictivos que evaluén distintas características de los jugadores; todo con el fin de tomar decisiones concretas basadas en los datos.

Sin embargo, su condición de corporeo le impidió tomar la versión anterior de MDS7202, por lo que este motivo Renacín contrata a su equipo para lograr su objetivo final. Dado que aún tiene fuertes vínculos con la dirección de deportes de la municipalidad, el corporeo le entrega base de datos con las estadísticas de cada jugador para que su equipo empieze a trabajar ya con un dataset listo para ser usado.


**Los Datos**

Para este laboratorio deberán trabajar con el csv `statsplayers.csv`, donde deberán aplicar algoritmos de aprendizaje supervisado de clasificación en base a características que describen de jugadores de fútbol.

Para comenzar cargue el dataset señalado y a continuación vea el reporte **`Player_Stats_Report.html`** (adjunto en la carpeta del enunciado) que describe las características principales del `DataFrame`.

In [1]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


In [2]:
import pandas as pd

df = pd.read_csv(r'.\stats_players.csv')

df


Name          Nationality National_Position Club_Position  \
0      Cristiano Ronaldo             Portugal                LS            LW   
1           Lionel Messi            Argentina                RW            RW   
2                 Neymar               Brazil                LW            LW   
3            Luis Suárez              Uruguay                LS            ST   
4           Manuel Neuer              Germany                GK            GK   
...                  ...                  ...               ...           ...   
17583        Adam Dunbar  Republic of Ireland               NaN           Sub   
17584       Dylan McGoey  Republic of Ireland               NaN           Sub   
17585    Tommy Ouldridge              England               NaN           Res   
17586         Mark Foden             Scotland               NaN           Sub   
17587   Barry Richardson              England               NaN           Sub   

       Height  Weight Preffered_Foot  Age        Work_Rate  Weak_foot  ...  \
0         185      80          Right   32       High / Low          4  ...   
1         170      72           Left   29  Medium / Medium          4  ...   
2         174      68          Right   25    High / Medium          5  ...   
3         182      85          Right   30    High / Medium          4  ...   
4         193      92          Right   31  Medium / Medium          4  ...   
...       ...     ...            ...  ...              ...        ...  ...   
17583     183      82          Right   19  Medium / Medium          1  ...   
17584     185      80          Right   19  Medium / Medium          2  ...   
17585     173      61          Right   18    High / Medium          2  ...   
17586     180      80          Right   21  Medium / Medium          3  ...   
17587     185      77          Right   47  Medium / Medium          2  ...   

       Agility  Jumping  Heading  Shot_Power  Finishing  Long_Shots  Curve  \
0           90       95       85          92         93          90     81   
1           90       68       71          85         95          88     89   
2           96       61       62          78         89          77     79   
3           86       69       77          87         94          86     86   
4           52       78       25          25         13          16     14   
...        ...      ...      ...         ...        ...         ...    ...   
17583       27       56       14          16         13          13     11   
17584       28       53       12          17         12          11     12   
17585       54       61       41          44         28          42     35   
17586       34       48       15          23         14          12     13   
17587       38       51       12          13         11          16     12   

       Freekick_Accuracy  Penalties  Volleys  
0                     76         85       88  
1                     90         74       85  
2                     84         81       83  
3                     84         85       88  
4                     11         47       11  
...                  ...        ...      ...  
17583                 13         15       12  
17584                 13         16       12  
17585                 36         42       37  
17586                 12         24       12  
17587                 11         22       12  

[17588 rows x 39 columns]

In [3]:
df.shape

(17588, 39)

In [4]:
df.dtypes


Name                 object
Nationality          object
National_Position    object
Club_Position        object
Height                int64
Weight                int64
Preffered_Foot       object
Age                   int64
Work_Rate            object
Weak_foot             int64
Skill_Moves           int64
Ball_Control          int64
Dribbling             int64
Marking               int64
Sliding_Tackle        int64
Standing_Tackle       int64
Aggression            int64
Reactions             int64
Interceptions         int64
Vision                int64
Composure             int64
Crossing              int64
Short_Pass            int64
Long_Pass             int64
Acceleration          int64
Speed                 int64
Stamina               int64
Strength              int64
Balance               int64
Agility               int64
Jumping               int64
Heading               int64
Shot_Power            int64
Finishing             int64
Long_Shots            int64
Curve               

## 1. Predicción de Seleccionados Nacionales [14 puntos]

### 1.1 Preprocesamiento [5 puntos]

Tareas:

1. Genere los labels para la clasificación binaria en una variable llamada `label`. Para esto, trabaje sobre el atributo `National_Position` suponiendo que los valores nulos son jugadores no seleccionados para representar a su país. [Sin puntaje]

2. Hecho esto, ¿cuántos se tienen ejemplos por cada clase? Comente lo que observa. [1 punto]

3. Genere un `ColumnTransformer` en donde especifique las transformaciones que hay que realizar para cada columna (por ejemplo StandarScaler, MinMaxScaler, OneHotEncoder, etc...) para que puedan ser utilizadas correctamente por el modelo predictivo y guárdelo una variable llamada `col_transformer`. [2 puntos]

4. Comente y justifique las transformaciones elegidas sobre cada una de las variables (para esto utilice el material `Player_Stats_Report.html` que viene en el zip del lab), al igual que las transformaciones aplicadas. [2 puntos]

**Respuesta:**

In [5]:
df['label'] = df['National_Position'].notnull().astype(int)

class_counts = df['label'].value_counts()
print(class_counts)

label
0    16513
1     1075
Name: count, dtype: int64


Hay 1075 jugadores seleccionados nacionales y 16513 jugadores no seleccionados nacionales.

In [6]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, MinMaxScaler
from sklearn.pipeline import Pipeline


numeric_var = df.select_dtypes(include=['int64']).columns.tolist()
categorical_var = df.select_dtypes(include=['object']).columns.tolist()

col_transformer = Pipeline([
    ('col_transformer', ColumnTransformer([
            ('int', MinMaxScaler(), numeric_var),
            ('object', OneHotEncoder(handle_unknown='ignore'), categorical_var)]
    ))
])

col_transformer

Pipeline(steps=[('col_transformer',
                 ColumnTransformer(transformers=[('int', MinMaxScaler(),
                                                  ['Height', 'Weight', 'Age',
                                                   'Weak_foot', 'Skill_Moves',
                                                   'Ball_Control', 'Dribbling',
                                                   'Marking', 'Sliding_Tackle',
                                                   'Standing_Tackle',
                                                   'Aggression', 'Reactions',
                                                   'Interceptions', 'Vision',
                                                   'Composure', 'Crossing',
                                                   'Short_Pass', 'Long_Pass',
                                                   'Acceleration', 'Speed',
                                                   'Stamina', 'Strength',
                                                   'Balance', 'Agility',
                                                   'Jumping', 'Heading',
                                                   'Shot_Power', 'Finishing',
                                                   'Long_Shots', 'Curve', ...]),
                                                 ('object',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['Name', 'Nationality',
                                                   'National_Position',
                                                   'Club_Position',
                                                   'Preffered_Foot',
                                                   'Work_Rate'])]))])

Se decide realizar transformaciones específicas tanto para las variables numéricas como para las categóricas con el fin de preparar el conjunto de datos para su uso en modelos de futbol que se vienen.

Para las variables numéricas, se utilizó la transformación `StandardScaler`. Esta elección se fundamenta en la necesidad de estandarizar las variables que presentan diferentes escalas y rangos, tales como `Height`, `Weight`, `Age`, `Ball_Control`, `Dribbling`, entre otras. La estandarización mediante `StandardScaler` ajusta los datos para que tengan una media de 0 y una desviación estándar de 1, lo cual es crucial para algoritmos de aprendizaje automático que son sensibles a la escala de los datos.

En cuanto a las variables categóricas, se aplicó la transformación `OneHotEncoder`. Las variables categóricas como `Name`, `Nationality`, `National_Position`, `Club_Position`, `Preffered_Foot`, y `Work_Rate` fueron transformadas utilizando este método. `OneHotEncoder` convierte las categorías en una matriz binaria, permitiendo su uso en los modelos de futbol.


### 1.2 Entrenamiento [3 puntos]

Ahora, vamos a entrenar los pipelines generados en los pasos anteriores. Para esto, debe realizar las siguientes tareas:

1. Separe los datos de entrenamiento en un conjunto de entrenamiento y de prueba  (la proporción queda a su juicio). En este paso, seleccione los ejemplos de forma aleatoria e intente mantener la distribución original de labels de cada clase en los conjuntos de prueba/entrenamiento. (vea la documentación de `train_test_split`). [1 puntos]


2. Defina un pipeline llamado `pipeline_xgboost` y otro llamado `pipeline_lightgbm`. Estos pipelines deben tener el mismo ColumnTransformer definido en la sección de preprocesamiento, pero deben variar los clasificadores de acuerdo al nombre de cada pipeline. [1 puntos]

3. Entrene los pipelines. [1 punto]



In [7]:
import os
os.environ['LOKY_MAX_CPU_COUNT'] = '8'  


**Respuesta:**

In [8]:
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier

# 1
X = df.drop(columns=['label'])
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, stratify=y, random_state=42)

# 2
pipeline_xgboost = Pipeline([
    ('preprocessor', col_transformer),
    ('classifier', XGBClassifier())
])

pipeline_lightgbm = Pipeline([
    ('preprocessor', col_transformer),
    ('classifier', LGBMClassifier())
])

# 3
pipeline_xgboost.fit(X_train, y_train)
pipeline_lightgbm.fit(X_train, y_train)

[LightGBM] [Info] Number of positive: 752, number of negative: 11559
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.001215 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 2660
[LightGBM] [Info] Number of data points in the train set: 12311, number of used features: 149
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.061084 -> initscore=-2.732483
[LightGBM] [Info] Start training from score -2.732483
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain

Pipeline(steps=[('preprocessor',
                 Pipeline(steps=[('col_transformer',
                                  ColumnTransformer(transformers=[('int',
                                                                   MinMaxScaler(),
                                                                   ['Height',
                                                                    'Weight',
                                                                    'Age',
                                                                    'Weak_foot',
                                                                    'Skill_Moves',
                                                                    'Ball_Control',
                                                                    'Dribbling',
                                                                    'Marking',
                                                                    'Sliding_Tackle',
                                                                    'Standing_Tackle',
                                                                    'Aggression',
                                                                    'Reactions',
                                                                    'Interceptions',
                                                                    'Vision',
                                                                    'Composure',
                                                                    'Crossing',
                                                                    'Short_Pass',
                                                                    'Long_Pass',
                                                                    'Acceleration',
                                                                    'Speed',
                                                                    'Stamina',
                                                                    'Strength',
                                                                    'Balance',
                                                                    'Agility',
                                                                    'Jumping',
                                                                    'Heading',
                                                                    'Shot_Power',
                                                                    'Finishing',
                                                                    'Long_Shots',
                                                                    'Curve', ...]),
                                                                  ('object',
                                                                   OneHotEncoder(handle_unknown='ignore'),
                                                                   ['Name',
                                                                    'Nationality',
                                                                    'National_Position',
                                                                    'Club_Position',
                                                                    'Preffered_Foot',
                                                                    'Work_Rate'])]))])),
                ('classifier', LGBMClassifier())])

### 1.3 Resultados [6 puntos]

1. Calcule las métricas accuracy, precisión y recall de la clase positiva (la que indica que un jugador es seleccionado nacional) para evaluar el rendimiento de los distintos modelos. Verifique sus resultados usando `classification_report`. [2 puntos]

2. Explique qué implican los valores de accuracy, precisión y recall de la clase positiva (la que indica que un jugador es seleccionado nacional) y cómo influye la cantidad de ejemplos por clase en los resultados obtenidos. [2 puntos]

3. Explique qué métrica le parece más adecuada y concluya qué modelo tiene un mejor desempeño. [2 puntos]

In [9]:
from sklearn.metrics import classification_report, confusion_matrix

y_pred_xgboost = pipeline_xgboost.predict(X_test)
y_pred_lightgbm = pipeline_lightgbm.predict(X_test)

print("XGBoost Classification Report:")
print(classification_report(y_test, y_pred_xgboost, target_names=['No Seleccionado', 'Seleccionado']))

print("LightGBM Classification Report:")
print(classification_report(y_test, y_pred_lightgbm, target_names=['No Seleccionado', 'Seleccionado']))

print("XGBoost Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_xgboost))

print("LightGBM Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_lightgbm))

XGBoost Classification Report:
                 precision    recall  f1-score   support

No Seleccionado       1.00      1.00      1.00      4954
   Seleccionado       1.00      1.00      1.00       323

       accuracy                           1.00      5277
      macro avg       1.00      1.00      1.00      5277
   weighted avg       1.00      1.00      1.00      5277

LightGBM Classification Report:
                 precision    recall  f1-score   support

No Seleccionado       1.00      1.00      1.00      4954
   Seleccionado       1.00      1.00      1.00       323

       accuracy                           1.00      5277
      macro avg       1.00      1.00      1.00      5277
   weighted avg       1.00      1.00      1.00      5277

XGBoost Confusion Matrix:
[[4954    0]
 [   0  323]]
LightGBM Confusion Matrix:
[[4954    0]
 [   0  323]]


La métrica más adecuada en este contexto, donde el objetivo es clasificar jugadores seleccionados para un equipo nacional, es el recall de la clase positiva (jugadores seleccionados). Esto se debe a que, en situaciones donde una clase es minoritaria (323 seleccionados frente a 4954 no seleccionados), es crucial minimizar los falsos negativos, es decir, evitar que jugadores que deberían ser seleccionados no sean detectados por el modelo. En escenarios de selección, es más importante que el modelo no pierda a los jugadores valiosos que deben ser seleccionados, incluso si en el proceso se incluyen algunos jugadores que no deberían ser seleccionados (un problema que impacta más la precisión). En este caso, tanto el recall como la precisión son perfectos, lo que sugiere que el modelo no está fallando en identificar a ningún jugador seleccionado.

En cuanto a qué modelo tiene mejor desempeño, tanto XGBoost como LightGBM ofrecen métricas idénticas: precisión, recall y F1-score perfectos para ambas clases. Esto indica que ambos modelos manejan adecuadamente el desbalance de clases y predicen sin errores. Sin embargo, debido a que la situación es inusualmente perfecta, sería prudente cuestionar si estos resultados se mantienen bajo otras condiciones o configuraciones del modelo, ya que en un escenario real esta perfección sería improbable.


**Nota para ayudante:** Por favor entregar retroalimentación sobre las clases o si hubo algun error en el preprocesamiento, he estado mucho tiempo intentando balancear esto. gracias! :)

## 2. Predicción de posiciones de jugadores [4 puntos]

En una nueva jornada de desmesuradas transacciones deportivas, Renacín escuchó a sus colegas discutir acerca de que el precio de cada jugador depende en gran medida de la posición en la cancha en la que juega. Y además, que hay bastantes jugadores nuevos que no tienen muy claro en que posición verdaderamente brillarían, por lo que actualmente puede que actualmente estén jugando en posiciones sub-optimas.

Viendo que los resultados del primer análisis no son tan esperanzadores, el corporeo los comanda a cambiar su tarea: ahora, les solicita que construyan un clasificador enfocado en predecir la mejor posición de los jugadores en la cancha según sus características.

Para lograr esto, primero, les pide que etiqueten de la siguiente manera los valores que aparecen en el atributo `Club_Position`, pidiendo que agrupen los valores en los siguientes grupos:

**Nota**:  Renacín les recalca que **no deben utilizar los valores ```Sub``` y ```Res``` de esta columna**.

```python
ataque = ['ST', 'CF']
central_ataque = ['RW', 'CAM', 'LW']
central = ['RM', 'CM', 'LM']
central_defensa = ['RWB', 'CDM', 'LWB']
defensa = ['RB', 'CB', 'LB']
arquero = ['GK']
```

La elección del clasificador se justificar en base a la siguiente [guía](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html) y se deben comentar los resultados obtenidos en la clasificación.

**Tareas:** [1 punto por tarea]

1. En un nuevo dataframe, aplique las etiquetas descritas anteriormente en cada uno de los valores señalados en esta sección y guárdelos en la variable `label`.
2. Cuente cuántos por clase quedan.
3. Entrene el nuevo pipeline y ejecute una evaluación de este.  
4. Comente los resultados obtenidos.

**Respuesta:**

In [10]:
ataque = ['ST', 'CF']
central_ataque = ['RW', 'CAM', 'LW']
central = ['RM', 'CM', 'LM']
central_defensa = ['RWB', 'CDM', 'LWB']
defensa = ['RB', 'CB', 'LB']
arquero = ['GK']

def categorize_position(position):
    if position in ataque:
        return 'ataque'
    elif position in central_ataque:
        return 'central_ataque'
    elif position in central:
        return 'central'
    elif position in central_defensa:
        return 'central_defensa'
    elif position in defensa:
        return 'defensa'
    elif position in arquero:
        return 'arquero'
    else:
        return None
    
df['label'] = df['Club_Position'].apply(categorize_position)

In [11]:
print("Cantidad de valores por clase:\n",df['label'].value_counts())
# print(df.shape)

Cantidad de valores por clase:
 label
defensa            1180
central             907
arquero             632
central_ataque      581
ataque              430
central_defensa     209
Name: count, dtype: int64


In [12]:
# START
# > 50+ samples (yes)
# > predicting a category (yes)
# > do you have labeled data (yes)
# > 100k+ samples (no)
# > SGD Classifier

from sklearn.linear_model import SGDClassifier
from sklearn.metrics import classification_report, confusion_matrix

print(df.shape)
df = df.dropna()
print(df.shape)

X = df.drop(columns=['label'])
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, stratify=y, random_state=42)

pipeline_sgdclassifier = Pipeline([
    ('preprocessor', col_transformer),
    ('classifier',SGDClassifier())
])

pipeline_sgdclassifier.fit(X_train, y_train)
y_pred = pipeline_sgdclassifier.predict(X_test)

eval = classification_report(y_test, y_pred) # classification report
print(eval)

c_matrix = confusion_matrix(y_test, y_pred) # confusion matrix
print(c_matrix)

(17588, 40)
(366, 40)
                 precision    recall  f1-score   support

        arquero       1.00      1.00      1.00        24
         ataque       1.00      1.00      1.00        13
        central       1.00      0.95      0.98        21
 central_ataque       1.00      1.00      1.00        20
central_defensa       1.00      0.50      0.67         6
        defensa       0.87      1.00      0.93        26

       accuracy                           0.96       110
      macro avg       0.98      0.91      0.93       110
   weighted avg       0.97      0.96      0.96       110

[[24  0  0  0  0  0]
 [ 0 13  0  0  0  0]
 [ 0  0 20  0  0  1]
 [ 0  0  0 20  0  0]
 [ 0  0  0  0  3  3]
 [ 0  0  0  0  0 26]]


## 3. Predicciones de Seleccionados Nacionales para el Jere Klein [30 puntos]

Después de alcanzar la fama como cantante urbano, Jere Klein decide explorar una nueva faceta. Con su amor por el fútbol y convencido de que los artistas urbanos poseen un talento y versatilidad excepcionales, Jere se embarca en un proyecto innovador: desarrollar un sistema de inteligencia artificial capaz de identificar a jugadores que tienen potencial para convertirse en futbolistas profesionales. Su teoría es que muchos artistas del género urbano chileno, con sus habilidades únicas y su disciplina, podrían destacarse también en el deporte. Con este sistema, Jere espera no solo abrir nuevas oportunidades para sus colegas artistas, sino también demostrar la amplia gama de talentos que pueden ofrecer.

### 2.1 ¿Qué modelo de árbol es más de "pana"? [10 puntos]


**Tareas**


1. Considerando el la variable llamada `label` creada en la sección 1.1. Para determinar cuál modelo de árbol sería más adecuado para la tarea en cuestión, utilice PyCaret. Este deberá centrarse exclusivamente en modelos de tipo árbol. Jere ha especificado que busca un modelo que tome decisiones rápidamente y que tenga una baja tasa de falsos positivos, ya que planea invertir en estos jugadores. [3 puntos] 

Para la comparación, utilice los siguientes modelos:

```python
['et', 'rf', 'dt', 'xgboost', 'lightgbm', 'catboost']
```

2. Explique en brevemente que son los modelos de la siguiente lista `['et', 'rf', 'dt']` y como funcionan. [3 punto]

3. Tras realizar la comparación de modelos, seleccione aquel que muestre el mejor rendimiento en términos de velocidad y precisión, especialmente en la reducción de falsos positivos. Utilice la función `evaluate_model` de PyCaret para revisar y analizar los resultados obtenidos en los siguientes aspectos:

  - **Confusión Matrix**: ¿Cómo se encuentran la tasa de verdaderos positivos y verdaderos negativos?
  - **Threshold**: ¿Es acaso el umbral por defecto del modelo el mejor para las predicciones?
  - **Feature Importance**: ¿Cuáles son las variables con mejor desempeño? ¿A qué podría deberse esto?
  - **Learning Curve**: ¿El modelo presenta algún problema?

  [4 puntos]

**Respuesta**

In [13]:
from pycaret.datasets import get_data
from pycaret.classification import setup, compare_models, evaluate_model
import os

os.environ["PYCARET_CUSTOM_LOGGING_LEVEL"] = "CRITICAL"

# #Continuar código aquí

# as in 1.1
df = pd.read_csv(r'./stats_players.csv')
df['label'] = df['National_Position'].notnull().astype(int)

# setup is used from hereon
clf_setup = setup(
    data=df, 
    target='label', 
    session_id=42, 
    use_gpu=True
)

model_list = ['et', 'rf', 'dt', 'xgboost', 'lightgbm', 'catboost']

# compare
print('compare_models')
best_model = compare_models(include=model_list)

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] Number of positive: 1, number of negative: 1
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4070 Ti SUPER, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] Number of positive: 1, number of negative: 1
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4070 Ti SUPER, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves

**'dt' (DecisionTreeClassifier):**
El Árbol de Decisión se refiere a un algoritmo supervisado que divide progresivamente los datos en función de los valores de las características que proporcionan la mejor separación entre categorías. El proceso continúa hasta que todos los datos se hayan dividido o hasta que se alcance algún umbral. Los Árboles de Decisión son propensos a sobreajustarse, ya que encuentran la mejor separación para el conjunto de datos dado, sin preocuparse por la capacidad de generalización.

**'rf' (RandomForestClassifier):**
Random Forest es un método de conjunto que construye múltiples árboles de decisión y combina sus resultados para crear un modelo más robusto. Los árboles se construyen a partir de una muestra aleatoria de puntos de datos y se evalúan con una selección aleatoria de características. En cada paso, se elige el criterio que mejor divide la muestra. El proceso se detiene cuando se cumple un criterio dado, y luego se elige el criterio de mayoría (para clasificación) o promedio (para regresión). La aleatoriedad introducida por este proceso ayuda a evitar el sobreajuste en la categorización.

**'et' (ExtraTreesClassifier):**
Extra Trees, al igual que Random Forest, construye múltiples árboles de decisión en diferentes subconjuntos de datos, pero introduce aún más aleatoriedad al elegir de forma aleatoria los puntos de división en lugar de usar el mejor. Esta capa extra de aleatoriedad y su criterio más flexible hacen que Extra Trees sea menos propenso al sobreajuste que Random Forest, además de ser menos intensivo computacionalmente, ya que no se optimizan los puntos de división.

### 2.2 Reducción de dimensionalidad [14 puntos]

A pesar de los resultados obtenidos previamente, el manager de Jere ha solicitado el entrenamiento de un modelo de XGBoost utilizando los datos disponibles. Además, se debe proceder a realizar una reducción de dimensionalidad basada en la importancia de las características.

Para llevar a cabo esta tarea:

1. Inicie entrenando un modelo XGBoost con todas las características disponibles. [2 puntos]

2. Una vez el modelo esté entrenado, evalúe y clasifique las características según su importancia de forma descendente. [2 puntos]

3. Utilice esta clasificación para ejecutar una búsqueda recursiva de eliminación de características, eliminando progresivamente las menos importantes y evaluando el impacto en el desempeño del modelo hasta identificar las N características más críticas. [2 puntos]

4. Con este conjunto reducido de características, entrene un nuevo modelo y evalúe su rendimiento. [2 puntos]

5. Posteriormente, responda a las siguientes preguntas para una comprensión más profunda de los cambios y beneficios:

  - ¿El rendimiento del modelo con las características seleccionadas es similar al del modelo original? ¿Cómo se comparan en términos de precisión y robustez? [2 puntos]
  - ¿Cuáles son los beneficios potenciales de eliminar variables del modelo? Considere factores como la simplificación del modelo, reducción del tiempo de entrenamiento, y mejora en la capacidad de generalización. [2 puntos]
  - Comente si el modelo con menor dimensionalidad es más sencillo de explicar. Explique brevemente por qué la eliminación de ciertas características puede facilitar la comprensión y la explicación del comportamiento del modelo. [2 puntos]

Notar que con esta metodologia buscamos encontrar un punto entermedio entre número de festures y desempeño. por esto, si observa que al aumentar festires el aumento es despreciable, puede no considerar agregar más features a su modelo.

**Respuesta**

*"Inicie entrenando un modelo XGBoost con todas las características disponibles."*

Por alguna razón, el código deja de correr en la siguiente celda. No me deja intentarlo de nuevo. Esto pasó de repente, y en teoría debería correr por sí solo. Lo siento.

In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import pandas as pd

# as 1.1
df = pd.read_csv(r'./stats_players.csv')
df['label'] = df['National_Position'].notnull().astype(int)

# as before
numeric_var = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_var = df.select_dtypes(include=['object']).columns.tolist()
categorical_var.remove('National_Position')

# as before
col_transformer = ColumnTransformer(transformers=[
    ('num', MinMaxScaler(), numeric_var),
    ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_var)
])

# def X,y
X = df.drop(columns=['label', 'National_Position'])
y = df['label']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, stratify=y, random_state=42)

# as before
pipeline_xgboost = Pipeline([
    ('preprocessor', col_transformer),
    ('classifier', XGBClassifier(tree_method='gpu_hist', predictor='gpu_predictor')) # use GPU
])

# re train
pipeline_xgboost.fit(X_train, y_train)

# evaluate
y_pred = pipeline_xgboost.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)

print(f'Accuracy with XGBoost: {accuracy:.4f}')


*"Una vez el modelo esté entrenado, evalúe y clasifique las características según su importancia de forma descendente."*

In [ ]:
xgb_model = pipeline_xgboost.named_steps['classifier']

feature_importances = xgb_model.feature_importances_

feature_names = numeric_var + list(pipeline_xgboost.named_steps['preprocessor'].transformers_[1][1].get_feature_names_out(categorical_var))

# Create a DataFrame for feature importance and sort in descending order
importance_df = pd.DataFrame({
    'Característica': feature_names,
    'Importancia': feature_importances
}).sort_values(by='Importancia', ascending=False)

# Display the sorted feature importance
print(f'Ranking de características según importancia:\n{importance_df}')


*"Utilice esta clasificación para ejecutar una búsqueda recursiva de eliminación de características, eliminando progresivamente las menos importantes y evaluando el impacto en el desempeño del modelo hasta identificar las N características más críticas."*

In [ ]:
from sklearn.feature_selection import RFE

print("Características con importancia mayor que cero:",importance_df[importance_df['Importancia']>0].shape[0])

# set up RFE
rfe = RFE(estimator=pipeline_xgboost.named_steps['classifier'], n_features_to_select=10) # selecting 10 features because ben 10 is my muse
X_train_transformed = pipeline_xgboost.named_steps['preprocessor'].transform(X_train) # don't need to fit only hush hush
rfe.fit(X_train_transformed, y_train)

rankings = rfe.ranking_ # rank of importance, numeric
support = rfe.support_ # True if selected, bool

all_feature_names = numeric_var + list(onehot_var)  # lists of numeric and oneshotted variable names
selected_features = [feature for feature, rank in zip(all_feature_names, support) if rank == True] # feature has rank False if dropped

print("Características importantes:", selected_features)


*"Con este conjunto reducido de características, entrene un nuevo modelo y evalúe su rendimiento."*

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

X_train_selected = X_train_transformed[:, support] # all where rank rings true

# deal with the test transformation since i hadn't done it yet:
X_test_transformed = pipeline_xgboost.named_steps['preprocessor'].transform(X_test) # no need to fit just yet
X_test_selected = X_test_transformed[:, support] # ñ

# model 
model = XGBClassifier() # model
model.fit(X_train_selected, y_train) # model
# model# model
y_pred_selected = model.predict(X_test_selected)

accuracy_selected = accuracy_score(y_test, y_pred_selected)
conf_matrix_selected = confusion_matrix(y_test, y_pred_selected)
class_report_selected = classification_report(y_test, y_pred_selected)

print(f"Accuracy: {accuracy_selected:.2f}")
print("\nConfusion Matrix:")
print(conf_matrix_selected)
print("\nClassification Report:")
print(class_report_selected)

"*Posteriormente, responda a las siguientes preguntas para una comprensión más profunda de los cambios y beneficios:*

  - *¿El rendimiento del modelo con las características seleccionadas es similar al del modelo original? ¿Cómo se comparan en términos de precisión y robustez?*
  - *¿Cuáles son los beneficios potenciales de eliminar variables del modelo? Considere factores como la simplificación del modelo, reducción del tiempo de entrenamiento, y mejora en la capacidad de generalización.*
  - *Comente si el modelo con menor dimensionalidad es más sencillo de explicar. Explique brevemente por qué la eliminación de ciertas características puede facilitar la comprensión y la explicación del comportamiento del modelo.*"

### 2.3 Calibración Probabilistica [6 puntos]

Para lograr modelos más modulares, se recomienda realizar una calibración del modelo entrenado anteriormente, con el objetivo de obtener salidas que reflejen mayor modularidad.

1. Se solicita que utilice un método de calibración que asegure que las probabilidades generadas incrementen de manera monótona. Una métrica ampliamente utilizada para evaluar la precisión de la calibración de un modelo es el Brier Score. Calcule el Brier Score para el modelo tanto antes como después de la calibración. Esto le permitirá realizar una comparación cuantitativa y determinar si la calibración ha mejorado el rendimiento del modelo. Para más información sobre el Brier Score, puede consultar el siguiente enlace: [Scikit-Learn - Brier Score Loss](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.brier_score_loss.html). [3 puntos]

2. Tras la calibración, examine y comente los resultados obtenidos. A su análisis añada una comparación visual de las ideales versus las salidas del modelo original (sin calibrar) y del modelo calibrado. [3 puntos]

**Respuesta**

In [ ]:
from sklearn.calibration import CalibratedClassifierCV
from sklearn.metrics import brier_score_loss
from xgboost import XGBClassifier

# model
model = XGBClassifier() # model
model.fit(X_train_selected, y_train) # resultado de RFE

y_probs_before = model.predict_proba(X_test_selected)[:, 1] # i want the probabilities for class 1

# calibrate the pretrained model
calibrated_model = CalibratedClassifierCV(model, method='sigmoid', cv='prefit')
calibrated_model.fit(X_train_selected, y_train)

y_probs_after = calibrated_model.predict_proba(X_test_selected)[:, 1] # same as above but after cal

# get briar scores
brier_score_before = brier_score_loss(y_test, y_probs_before)
brier_score_after = brier_score_loss(y_test, y_probs_after)

print(f"Brier Score (before): {brier_score_before:.2f}")
print(f"Brier Score (after): {brier_score_after:.2f}")

In [ ]:
from sklearn.calibration import calibration_curve

ratio_pos_before, mean_pred_before = calibration_curve(y_test, y_probs_before, n_bins=10)
ratio_pos_after, mean_pred_after = calibration_curve(y_test, y_probs_after, n_bins=10)

plt.plot(ratio_pos_before, mean_pred_before, "s-", label="uncalibrated")
plt.plot(ratio_pos_after, mean_pred_after, "s-", label="calibrated")
plt.plot([0, 1], [0, 1], "k:", label="baseline")

plt.show()

In [24]:
# model